# Feature Scaliing

1.   Intro
2.   Import and describe the dataset
3.   Data preprocessing ( Train_test_split)  
4.   What happen on model without Scaling
5.   Logistic Regression without Scaling
6.   K-Nearest Neighbors without Scaling
7.   StandardScalar()
8.   Logistic Regression with StandardScalar
9.   K-Nearest Neighbors with StandardScalar
10.  MinMaxScalar()
11.  Logistic Regression with MinMaxScaler Explained
11.  K-Nearest Neighbors with MinMaxScaler Explained




## 2. Import and describe the dataset
For this topic, we use the Wine dataset of the scikit-learn which are already cleaned.

In [ ]:
import pandas as pd
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.linear_model import  LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

In [ ]:
wine = load_wine()
X, y = pd.DataFrame(wine.data, columns=wine.feature_names), wine.target

In [ ]:
X.head()

,alcohol,malic_acid,ash,alcalinity_of_ash,magnesium,total_phenols,flavanoids,nonflavanoid_phenols,proanthocyanins,color_intensity,hue,od280/od315_of_diluted_wines,proline
0,14.23,1.71,2.43,15.6,127.0,2.80,3.06,0.28,2.29,5.64,1.04,3.92,1065.0
1,13.20,1.78,2.14,11.2,100.0,2.65,2.76,0.26,1.28,4.38,1.05,3.40,1050.0
2,13.16,2.36,2.67,18.6,101.0,2.80,3.24,0.30,2.81,5.68,1.03,3.17,1185.0
3,14.37,1.95,2.50,16.8,113.0,3.85,3.49,0.24,2.18,7.80,0.86,3.45,1480.0
4,13.24,2.59,2.87,21.0,118.0,2.80,2.69,0.39,1.82,4.32,1.04,2.93,735.0


After checking `X.describe()`, we can see that the features in the wine dataset are on very different scales. For example, alcohol values are between about 11 and 15, while proline values range from 278 to 1680.

This big difference in scale can create problems for some machine learning algorithms, especially those that use distance `(like K-Nearest Neighbors)` or gradient descent `(like Logistic Regression)`. Features with larger numbers can have more influence on the model than features with smaller numbers.

To avoid this issue, we use feature scaling. Scaling makes all features have similar ranges so that each one contributes fairly to the model.

In [ ]:
X.describe()

,alcohol,malic_acid,ash,alcalinity_of_ash,magnesium,total_phenols,flavanoids,nonflavanoid_phenols,proanthocyanins,color_intensity,hue,od280/od315_of_diluted_wines,proline
count,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000
mean,13.000618,2.336348,2.366517,19.494944,99.741573,2.295112,2.029270,0.361854,1.590899,5.058090,0.957449,2.611685,746.893258
std,0.811827,1.117146,0.274344,3.339564,14.282484,0.625851,0.998859,0.124453,0.572359,2.318286,0.228572,0.709990,314.907474
min,11.030000,0.740000,1.360000,10.600000,70.000000,0.980000,0.340000,0.130000,0.410000,1.280000,0.480000,1.270000,278.000000
25%,12.362500,1.602500,2.210000,17.200000,88.000000,1.742500,1.205000,0.270000,1.250000,3.220000,0.782500,1.937500,500.500000
50%,13.050000,1.865000,2.360000,19.500000,98.000000,2.355000,2.135000,0.340000,1.555000,4.690000,0.965000,2.780000,673.500000
75%,13.677500,3.082500,2.557500,21.500000,107.000000,2.800000,2.875000,0.437500,1.950000,6.200000,1.120000,3.170000,985.000000
max,14.830000,5.800000,3.230000,30.000000,162.000000,3.880000,5.080000,0.660000,3.580000,13.000000,1.710000,4.000000,1680.000000


## 3. Train test split

`Train-Test Split` means dividing your data into two parts:

**Train data** → Used to teach the model.

**Test data** → Used to check how well the model learned.

*For example:*
If you have 100 rows of data,

80 rows → for training

20 rows → for testing

We do this because if we test the model on the same data it learned from, it may give very high accuracy.
Testing on new data shows how well the model works in real situations.

Train = Learn

Test = Check

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state= 42)

### 4. What Happens to a Model Without Scaling?

When we train a model without scaling:

- Features with **large values dominate** the model.
- Features with **small values get less importance**.
- Distance-based models (like KNN) perform poorly.
- Optimization-based models (like Logistic Regression) may show **convergence warnings** or train slowly.
- Accuracy can decrease or become misleading.

In short:  
**No scaling → Unbalanced learning → Poor or unreliable performance**

## 5. Logistic Regression Without Scaling

Logistic Regression is used for classification (not regression), meaning it predicts which class a data point belongs to.

###
Performance Without Scaling

When we trained the Logistic Regression model on unscaled data, we got an accuracy of 1.0 (100%). This looks perfect, but there’s an important issue.
```
ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
```

We received a ConvergenceWarning, which means the optimization algorithm (lbfgs) stopped because it reached the maximum number of iterations (max_iter = 2500) before properly finishing.

This usually happens when features have very different scales. Large differences in feature values make it harder for the algorithm to find the best solution efficiently.

So even though the accuracy is 100%:

The model may not have fully converged.

The parameters might not be truly optimal.

The high accuracy could just be because the dataset is easy to separate.

There’s also a possibility of overfitting.

That’s why scaling is still important, even if the accuracy looks perfect.

In [ ]:
lr = LogisticRegression(max_iter=2500)
lr.fit(X_train, y_train)

y_pred = lr.predict(X_test)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [ ]:
print("accuracy without scaling :")
accuracy_score(y_test, y_pred)

accuracy without scaling :


1.0

## 6. K-Nearest Neighbors (KNN) Without Scaling

K-Nearest Neighbors (KNN) is a simple classification algorithm.

It works by:
1. Choosing a value of **k** (for example, k = 5).
2. Finding the **k closest data points** to a new data point.
3. Assigning the class that appears most frequently among those neighbors.

KNN usually uses **Euclidean distance** to measure how close data points are.

### Performance Without Scaling

Without scaling, the KNN model achieved an accuracy of **0.72**, which is lower than Logistic Regression (1.0).

This is expected because KNN depends directly on distances, and different feature scales negatively affect its performance.

In short:

- **KNN + No Scaling → Lower Accuracy**
- **KNN + Scaling → Better Performance**

In [ ]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)


KNeighborsClassifier()

In [ ]:
knn_pred = knn.predict(X_test)

print("accuracy without scaling :")
accuracy_score(y_test, knn_pred)

accuracy without scaling :


0.7222222222222222

## 7. StandardScaler Explanation

`StandardScaler` is a preprocessing technique used to standardize features by removing the mean and scaling to unit variance. This process is often referred to as **Z-score normalization** or **standardization**.

### What it does:
For each feature (column) in the dataset, `StandardScaler` calculates the mean and standard deviation. Then, it transforms each data point `x` in that feature using the formula:

`z = (x - u) / s`

Where:
- `u` is the mean of the training samples.
- `s` is the standard deviation of the training samples

In [ ]:
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [ ]:
lr = LogisticRegression(max_iter=2500)
lr.fit(X_train_scaled, y_train)

LogisticRegression(max_iter=2500)

In [ ]:
y_scaled_pred = lr.predict(X_test_scaled)

In [ ]:
print("accuracy with scaling :")
accuracy_score(y_test, y_scaled_pred)

accuracy with scaling :


1.0

## 9. K-Nearest Neighbors (KNN) with StandardScaler

### What Happens When We Apply StandardScaler?

Applying `StandardScaler` greatly improves the performance of the KNN model, especially when features have different scales.

In our case:
- **KNN without scaling → Accuracy ≈ 0.72**
- **KNN with StandardScaler → Accuracy ≈ 0.94**

This is a significant improvement.

### What Does StandardScaler Do?

`StandardScaler` standardizes the data by:
- Subtracting the mean
- Dividing by the standard deviation  

After scaling:
- Each feature has **mean = 0**
- Each feature has **standard deviation = 1**

This makes all features have similar scales.

### Why Does This Improve KNN?

KNN is a **distance-based algorithm**. It calculates distances between data points to find the nearest neighbors.

Without scaling:
- Features with large values (like `proline`)
- Dominate the distance calculation
- Smaller-range features (like `alcohol`) have less influence

This leads to biased and incorrect neighbor selection.

With `StandardScaler`:
- All features contribute equally
- Distance calculations become balanced
- Nearest neighbors are identified more accurately


Scaling makes the distance calculation fair, which directly improves KNN performance.

In [ ]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train_scaled, y_train)

KNeighborsClassifier()

In [ ]:
knn_scaled_pred = knn.predict(X_test_scaled)

In [ ]:
print("accuracy with scaling :")
accuracy_score(y_test, knn_scaled_pred)

accuracy with scaling :


0.9444444444444444

## 10. MinMaxScaler
`MinMaxScaler` transforms features by scaling each feature to a given range, typically between 0 and 1. This scaling is done using the formula:

$$ X_{scaled} = \frac{X - X_{min}}{X_{max} - X_{min}} $$

Where:
- $X$ is the original feature value.
- $X_{min}$ is the minimum value of the feature.
- $X_{max}$ is the maximum value of the feature.


In [ ]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

## Logistic Regression with MinMaxScaler

### What Does MinMaxScaler Do?

`MinMaxScaler` scales all features to a fixed range, usually **0 to 1**.

It works by:
- Subtracting the minimum value
- Dividing by the range (max - min)

After scaling:
- The smallest value becomes **0**
- The largest value becomes **1**
- All other values fall between 0 and 1


### Why It Helps Logistic Regression

Logistic Regression uses optimization algorithms (like `lbfgs`) to find the best parameters.

If features are on very different scales:
- Training can be slow
- The model may struggle to converge
- Some features may dominate

With `MinMaxScaler`:
- All features are on the same scale
- Optimization becomes smoother
- Convergence improves
- The model becomes more stable


In [ ]:
lr_norm = LogisticRegression(max_iter=2500)
lr_norm.fit(X_train, y_train)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=2500)

In [ ]:
lr_norm_pred = lr_norm.predict(X_test)

print("accuracy with normalization :")
accuracy_score(y_test, lr_norm_pred)

accuracy with normalization :


1.0

## K-Nearest Neighbors with MinMaxScaler Explained

Applying `MinMaxScaler` significantly impacts the K-Nearest Neighbors (KNN) model's performance, especially when features have different scales. `MinMaxScaler` scales features to a fixed range, typically between 0 and 1, ensuring that all feature values fall within this defined interval.

In our analysis, the KNN model **without scaling** achieved an accuracy of approximately **0.7222** (as seen in cell `Ym2aJlcXM-rK`). However, after applying `MinMaxScaler` to the data, the KNN model's accuracy improved substantially to approximately **0.9444** (as seen in cell `rPiBIprENAID`).

Similar to `StandardScaler`, `MinMaxScaler` is crucial for distance-based algorithms like KNN. Without scaling, features with larger magnitudes (e.g., 'proline' ranging up to 1680.0) can dominate the distance calculations, making the model overly reliant on these features and overshadowing others with smaller ranges (e.g., 'alcohol'). By scaling all features to a common range (0-1), `MinMaxScaler` ensures that each feature contributes equally to the distance metric. This normalization leads to more accurate and meaningful distance computations, allowing the KNN algorithm to identify true nearest neighbors and consequently make better predictions, as demonstrated by the significant improvement in accuracy.

In [ ]:
knn_norm = KNeighborsClassifier(n_neighbors=5)

In [ ]:
knn_norm.fit(X_train_scaled, y_train)
knn_norm_pred = knn_norm.predict(X_test_scaled)
print("accuracy with normalization :")
accuracy_score(y_test, knn_norm_pred)

accuracy with normalization :


0.9444444444444444